crates/shim-kvm/src/allocator.rs

```diff
@@ -104,6 +106,244 @@ fn msb(val: usize) -> u32 {
+            return Err(AllocateError::NotAligned);
+        }
+
+        if size != align_down(size as _, Page::<Size4KiB>::SIZE) as usize {
+            return Err(AllocateError::NotAligned);
+        }
+
+        let curr_size = (2usize).checked_pow(msb(size)).unwrap();
+
+        let (first_half, first_half_size) = {
+            while self.allocator.free() < curr_size {
+                self.allocator.balloon();
+            }
+            let (chunk, chunk_size) = self.allocator.try_alloc_half(curr_size);
+
+            if chunk.is_null() {
+                self.allocator.balloon();
+                self.allocator.try_alloc_half(curr_size)
+            } else {
+                (chunk, chunk_size)
+            }
+        };
+
+        if first_half.is_null() {
+            return Err(AllocateError::OutOfMemory);
+        }
+
+        let second_half_size = size.checked_sub(first_half_size).unwrap();
+
+        if second_half_size > 0 {
+            if let Err(e) = self.allocate_and_map_memory(
+                map_to + first_half_size,
+                second_half_size,
+                flags,
+                parent_flags,
+            ) {
+                unsafe {
+                    self.allocator.dealloc_pages(first_half, first_half_size);
+                }
+                return Err(e);
+            }
+        }
+
+        let phys = shim_virt_to_enc_phys(first_half);
+        if let Err(e) = self.map_memory(phys, map_to, first_half_size, flags, parent_flags) {
+            unsafe {
+                self.allocator.dealloc_pages(first_half, first_half_size);
+            }
+            let _ = self.unmap_memory(map_to + first_half_size, second_half_size);
+            return Err(e);
+        }
+        // transmute the whole thing to one big bytes slice
+        Ok(unsafe { core::slice::from_raw_parts_mut(map_to.as_mut_ptr() as *mut u8, size) })
+    }
+
+    /// Map the zero page to the given virtual address
+    ///
+    /// FIXME: change PhysAddr to ShimPhysAddr to ensure encrypted memory
+    pub fn map_memory_zero(
+        &mut self,
+        map_to: VirtAddr,
+        size: usize,
+        mut flags: PageTableFlags,
+        parent_flags: PageTableFlags,
+    ) -> Result<&'static mut [u8], AllocateError> {
+        if size == 0 {
+            return Err(AllocateError::ZeroSize);
+        }
+
+        let page_range_to = {
+            let start = map_to;
+            let end = start + size - 1u64;
+            let start_page = Page::<Size4KiB>::containing_address(start);
+            let end_page = Page::<Size4KiB>::containing_address(end);
+            Page::range_inclusive(start_page, end_page)
+        };
+
+        if flags.contains(PageTableFlags::WRITABLE) {
+            flags.remove(PageTableFlags::WRITABLE);
+            flags |= PageTableFlags::BIT_10;
+        }
+
+        let zero_frame = *ZERO_PAGE_FRAME;
+
+        for page_to in page_range_to {
+            unsafe {
```

Comment:
Safety comment, and can we make this block smaller?

Reply:
done